### ISSUE TITLE
Shape error in ForeachSOAP in 0.12.6 (REGRESSION from 0.9.0)

### ISSUE THREAD

**Issue by lukasschmit** [ClashLuke/HeavyBall#7]
using `heavyball==0.12.6`

```python
opt = ForeachSOAP(
    self.parameters(),
    lr=LR,
    betas=(0.95, 0.95),
    weight_decay=0.01,
    precondition_frequency=2,
)
```

causes

```
File "/usr/local/lib/python3.11/dist-packages/heavyball/utils.py", line 110, in exp_avg_sq_
10:28:40 torch._foreach_addcmul_(state, grad, grad, value=1 - beta2)
10:28:40RuntimeError: output with shape [8192, 1, 1536] doesn't match the broadcast shape [8192, 8192, 1536]
```

however exact same model/opt works fine in heavyball `heavyball==0.9.0`

**Comment by ClashLuke** [comment]
Can't reproduce the error with the information provided. Could you share a bit more?

I'd assume the parameter has the shape `[8192, 1, 1536]`. However, running the SOAP test with this shape doesn't cause any errors
https://github.com/ClashLuke/HeavyBall/blob/591dff38eaf25ff9b2292db40f4b80cbdf2a2019/test/test_merge.py#L20-L41

**Comment by fizzAI** [comment]
Having a similar error trying to use SOAP in my fork of NanoT5 (though its the one with a preconditioner schedule) https://github.com/fizzAI/nanoT5/tree/busted
```
Traceback (most recent call last):
  File "/root/nanoT5/nanoT5/main.py", line 69, in main
    train(
  File "/root/nanoT5/nanoT5/utils/train_utils.py", line 209, in train
    optimizer.step()
  File "/usr/local/lib/python3.10/dist-packages/accelerate/optimizer.py", line 171, in step
    self.optimizer.step(closure)
  File "/usr/local/lib/python3.10/dist-packages/torch/optim/lr_scheduler.py", line 137, in wrapper
    return func.__get__(opt, opt.__class__)(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/optim/optimizer.py", line 487, in wrapper
    out = func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/heavyball/precond_schedule_palm_foreach_soap.py", line 98, in step
    denom = exp_avg_sq_(exp_avg_sq, grad_projected, old_debiased2, group['eps'])
  File "/usr/local/lib/python3.10/dist-packages/heavyball/utils.py", line 110, in exp_avg_sq_
    torch._foreach_addcmul_(state, grad, grad, value=1 - beta2)
RuntimeError: output with shape [32128, 1, 768] doesn't match the broadcast shape [32128, 32128, 768]
```
Understandably not the smallest test case xD, but I hope it's useful

**Comment by fizzAI** [comment]
Can confirm 0.9.0 also fixes this for me

**Comment by ClashLuke** [comment]
Sorry to hear it's broken, but great to see there's an open reproduction!\
However, I can't seem to figure out what the parameter shape is here, as `768` and `SOAP` are nowhere to be found in the repo.

**Comment by ClashLuke** [comment]
Reproduced

**Comment by ClashLuke** [comment]
Fixed in `0.13.0`

**Comment by fizzAI** [comment]
🎉 Thanks! Great library btw :3
